### PR TITLE
Replace consensus `Encodable`/`Decodable` Psbt Serialization

### PR DESCRIPTION
**Pull request by DanGould**
fix https://github.com/rust-bitcoin/rust-bitcoin/issues/934

Instead of using consensus {de,en}code, serialize high-level structures (PartiallySignedTransaciton, Output, Input) borrow the signature from `Serialize`, `Deserialize` traits and implement them on Psbt:

```rs
impl Psbt {
    /// Serialize a value as raw data.
    fn serialize(&self) -> Vec<u8>;

    /// Deserialize a value from raw data.
    fn deserialize(bytes: &[u8]) -> Result<Self, encode::Error>;
}
```

### CODE REVIEW

**Comment by tcharding** [comment]
Side note: If you use one of the nine keywords [defined by github](https://docs.github.com/en/enterprise-server@3.4/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) the issue will be automatically closed when a PR merges. 'Addresses' isn't on the list, or was that intentional?

File: src/util/psbt/macros.rs
Comment (tcharding) [inline src/util/psbt/macros.rs:19]: Just a personal tip: it makes review easier if the diff is clean (doesn't have stuff like this in it), anything you can do to make review easier helps you keep reviewers happy which helps you get your stuff reviewed.

File: src/util/psbt/mod.rs
Comment (tcharding) [inline src/util/psbt/mod.rs:54]: I know its a draft but there is too much going on in this PR for me to review it by diff, I had to check out the branch and play around with it (could just be me). The serde stuff could be done in a separate patch (even done later once you get the main bit working).

File: src/util/psbt/mod.rs
Comment (tcharding) [inline src/util/psbt/mod.rs:424]: A more natural call would be: `let ser = expected.serialize()`

File: src/util/psbt/serialize.rs
Comment (tcharding) [inline src/util/psbt/serialize.rs:402]: There is an off-by-one error here, should be a 5 - 4 bytes of magic, 1 byte of 0xff

**Review by tcharding — COMMENTED** [review]
I left some general contribution suggestions. As far as the code goes, I played around with it locally and the code more or less works AFAICT, once the off-by-one bug is fixed. I reverted the `hex_psbt` macro changes too and just changed the test `serialize_and_deserialize_global` to use the new ser/deser code.

One note, either the traits `Serialize` and `Deserialize` have to have there docs changed or else PSBT shouldn't implement them because doing so conflicts with their documentation. I kind of prefer to just implement two methods on PSBT, not sure how others feel about this. Also its all a bit of a tangle so there is probable more than one way to go forward and I do not have a picture in mind of the final destination :)

File: src/util/psbt/mod.rs
Comment (DanGould) [reply src/util/psbt/mod.rs]: Much of my confusion comes from incomplete understanding of the role of serde. Don't we have 2 `Serialize` Traits on PartiallySignedTransaction with this line?

File: src/util/psbt/macros.rs
Comment (DanGould) [reply src/util/psbt/macros.rs]: Thank you for excusing the mess. Won't happen again.

Comment (DanGould) [reply src/util/psbt/macros.rs]: > I reverted the hex_psbt macro changes too and just changed the test serialize_and_deserialize_global to use the new ser/deser code.

Shouldn't the hex_psbt macro serialize any psbt from hex? After fixing off by 1, tests still fail for e.g. InvalidMagic. Doesn't that mean there's an issue with the deserialze method? At first I thought the macro may deserialize any psbt-related piece of data with the consensus-encoding, but it looks like it's only deserializing Psbts. And this code behaves in ways the tests do not expect.

```
---- util::psbt::tests::bip_371_vectors::valid_psbt_vectors stdout ----
thread 'util::psbt::tests::bip_371_vectors::valid_psbt_vectors' panicked at 'called `Result::unwrap()` on an `Err` value: Psbt(InvalidMagic)', src/util/psbt/mod.rs:897:626
```

File: src/util/psbt/mod.rs
Comment (tcharding) [reply src/util/psbt/mod.rs]: You can implement multiple traits by the same name AFAIK, but like I say somewhere else, I don't think we should implement our  `Serialize` trait anyways.

File: src/util/psbt/macros.rs
Comment (tcharding) [reply src/util/psbt/macros.rs]: Looks like the cause of the failing tests is serializing with `Encodable` and trying to deserialize with `Deserialize`.

Comment (DanGould) [reply src/util/psbt/macros.rs]: Doesn't that mean one of the two is non-standard? Bip 174 defines a byte sequence on the wire

**Comment by DanGould** [comment]
@tcharding as you note the Serialize Deserialize Psbt Traits are for the key value pairs. I propose we continue to use the consensus_encoding signature with the io::Read and io::Write signature but in non-consensus-related methods. It's more convenient to use a decoder with a pointer than raw byte streams.

**Comment by DanGould** [comment]
I've taken a few attempts to solve this after the draft. I think the original reason Carl used consensus encoding is because that module contains convenient, generalized, and now well reviewed serialization patterns. The PSBT-pr reviewers knew the patterns and so ignored their repurpose. We should separate these patterns from the consensus module for their utility across the library. Then the consensus mod will contain consensus layer bitcoin p2p network implementation and a clean serialize mod will contain familiar utilities available for clean application level support for psbt and beyond.

**Comment by tcharding** [comment]
Isn't the problem we are trying to solve that implementing `Encodable` for PSBT is a strange thing to do. It does not matter _how_ we serialize stuff in the `psbt` module, if we choose to use `consensus_encode` under the hood thats fine (as we do for all the other implementations) since the exact method of serialization is up to us (no spec), right? The issue that `consensus_encode` accepts a writer may be a better API than `psbt::serialize::Serialize` is a separate issue as is the fact that we have our own `Serialize` trait as well as deriving `serde::Serialize`.

**Comment by DanGould** [comment]
Ah I think you cracked it. The simple solution wins again.

Makes sense to take advantage of `consensus_encode` `consensus_decode` under the hood without exposing the `Encodable` `Decodable` Traits.

> One note, either the traits Serialize and Deserialize have to have there docs changed or else PSBT shouldn't implement them because doing so conflicts with their documentation. I kind of prefer to just implement two methods on PSBT, not sure how others feel about this. Also its all a bit of a tangle so there is probable more than one way to go forward and I do not have a picture in mind of the final destination :)


What's wrong with the doc:
```
//! PSBT serialization.
//!
//! Defines traits used for (de)serializing PSBT values into/from raw
//! bytes from/as PSBT key-value pairs.
```
The [spec](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#specification) includes the whole \<psbt\> as a key-value pair itself

We can implement two methods on PSBT but should also implement them for every non-Consensus PSBT related data that implements consensus de/encode now to fix the kernel of the issue

File: src/util/psbt/mod.rs
Comment (sanket1729) [inline src/util/psbt/mod.rs:290]: We should also remove Encodable from here

**Review by sanket1729 — COMMENTED** [review]
concept ACK. This will be a breaking change. 

We should only have these methods as direct `impl Psbt` methods and make the psbt::Serialize trait private. Moving this discussion to the issue

**Comment by DanGould** [comment]
The bip 174 `<keypair> `is defined as `<keypair> := <key> <valuedata>` such that

```
 <keypair> := <key> <value>
 <key> := <keylen> <keytype> <keydata>
 <value> := <valuelen> <valuedata>
```

Our `Pair` struct forgoes <valuelen> because <value> is saved as Vec<u8>, which manages len itself.

https://github.com/rust-bitcoin/rust-bitcoin/blob/5b707069e570bb087d768808d5ea724d67e448c6/src/util/psbt/raw.rs#L31-L41

Instead `Pair` depends consensus `impl Encodable for Vec<u8>` which defines consensus_encode as consensus_encode_with_size, serializing the length as defined in spec.

**Is this desirable since Vec<u8> already manages `len`?** Or should our internal representation be a 1:1 map of the spec struct to memory? What mental model makes the most sense?

This snagged me when I took `buf.extend(Pair.value: Vec<u8>)` granted for <value> and not as just <valuedata>.

**Comment by DanGould** [comment]
Another major question I have is whether we want to keep the `serialize(&self) -> Vec<u8>` signatuer, great for in memory operations, or `consensus_encode(&self, &mut w) -> Result<usize, io::Error>`  with results. The serialize is nice because we don't have to account for whole categories of errors, while the latter is nice to write to other mediums and return errors from things like the file system if a disk were ejected.

**Comment by apoelstra** [comment]
If we impl `consensus_encode` then we get [`serialize()` for free](https://docs.rs/bitcoin/latest/bitcoin/consensus/encode/fn.serialize.html).

**Comment by DanGould** [comment]
> If we impl `consensus_encode` then we get [`serialize()` for free](https://docs.rs/bitcoin/latest/bitcoin/consensus/encode/fn.serialize.html).

@apoelstra now I'm confused. Isn't this issue trying to fix that all these application layer messages use `impl Encodable`/`impl Encodable` but should not? re: #934

**Comment by Kixunil** [comment]
I'd love to have a `genio`-style trait with associated `Error`. If this is considered too much then probably impl for writer and unwrap the result in `fn serialize`.

**Comment by apoelstra** [comment]
> > If we impl `consensus_encode` then we get [`serialize()` for free](https://docs.rs/bitcoin/latest/bitcoin/consensus/encode/fn.serialize.html).
> 
> @apoelstra now I'm confused. Isn't this issue trying to fix that all these application layer messages use `impl Encodable`/`impl Decodable` but should not? re: #934

Ah, yeah, I see. That's pretty annoying. To answer your original question yes, I woud like to keep the `serialize(&self) -> Vec<u8>` siganture.

**Review by Kixunil — CHANGES_REQUESTED** [review]
This should remove the `Psbt` variant from `consensus::encode::Error`. Also the tests should start passing after rebasing.

What else is blocking this?

**Comment by DanGould** [comment]
This should work end to end now. I welcome feedback on the design. I'm not sure why the fuzz test fails yet.

@Kixunil I agree this Pr is prerequisite to fixing #837, but I think the fix another Pr. The `consensus` / `psbt` untangling looks more involved than it appears on the surface in my brief attempt. One thing would be to bring over an `io::Error`  psbt::Error which is somewhat involved since io::Error breaks `#[derive(Eq)]` which is required in tests. My attempt building on this pr is [on this branch](https://github.com/DanGould/rust-bitcoin/tree/psbt-ser-error)

**Comment by Kixunil** [comment]
@DanGould I believe that the best error untangling would be to make decodable error associated.

Regarding fuzzing, weird I don't see the reproducer in the output.

File: bitcoin/src/psbt/serialize.rs
Comment (Kixunil) [inline bitcoin/src/psbt/serialize.rs:67]: This worsens performance quite a bit. But maybe OK as a first iteration and it can be fixed later?

**Review by Kixunil — COMMENTED** [review]
Just skimmed it, I don't have much time. Looks quite nice, just the performance impact is bugging me. (But there was already shitton of allocations so it wasn't suitable for high-perf anyway.)

File: bitcoin/src/psbt/serialize.rs
Comment (DanGould) [reply bitcoin/src/psbt/serialize.rs:67]: What resources would you recommend to pursue a more performant iteration?

Comment (Kixunil) [reply bitcoin/src/psbt/serialize.rs:67]: You could serialize into `io::Write`r to avoid allocations. But really, we should have our own trait similar to [the one in `genio`](https://docs.rs/genio/latest/genio/trait.Write.html) to avoid having to depend on `core2` and to handle errors properly.

**Comment by DanGould** [comment]
> I believe that the best error untangling would be to make decodable error associated. -- @Kixunil 

I took a stab keeping the `io::Error` as associated `encode::Error` in [this commit](https://github.com/DanGould/rust-bitcoin/commit/f47efcbec5437b7b17d6eedc69a65fb29908ca7d). It works.

I wonder however, is a problem to instantiate `consensus::encode::Error` in this module for `io::Error`? This instance does not come from `consensus` module behavior? I would not consider this solution a regression since we were already using `encode::Error` in this case before, but it rhymes with the design problem this PR intends to fix.

**Comment by Kixunil** [comment]
I meant having associated `Error` type at `Decodable` trait. Encoding should only fail if the writer fails so we should have `io::Error` there (at least until we have something like `genio`).

> is a problem to instantiate consensus::encode::Error in this module for `io::Error`?

It certainly does look strange. Maybe we should have this shared type:

```rust
/// Error used when reading or decoding fails.
pub enum ReadError<Io, Decode> {
    /// Reading failed
    Io(Io),
    /// Decoding failed
    Decode(Decode), // consensus and PSBT error here
}
```

However this one will be annoying to use with `?` :( We could have `ResultExt` to provide `decode()` and `io()` methods to make it easier.

If that's not acceptable then I think deduplicated IO error is better.

**Comment by DanGould** [comment]
@tcharding @sanket1729 would you please take a look at this? This should be a complete, working replacement.

**Review by tcharding — DISMISSED** [review]
ACK e9b058d18efd48d4203375712c575905c0cdd389

**Comment by tcharding** [comment]
I don't think the error stuff is perfect but this PR looks like progress to me. Thanks @DanGould!

**Comment by DanGould** [comment]
I plan to fix error stuff in the next PR. Now we have a way for it to be fixed 🫡

**Comment by tcharding** [comment]
I believe your fuzz CI fail is because there are a bunch of funny files in `hfuzz_workspace/deserialize_psbt`.

**Comment by DanGould** [comment]
@tcharding are you implying this is an issue with the code, the test, or the test environment with hfuzz_workspace? I'm guessing this a caching CI environment issue. hfuzz_workspace does not look like a source file.

File: bitcoin/src/psbt/serialize.rs
Comment (sanket1729) [inline bitcoin/src/psbt/serialize.rs:71]: @DanGould, the fuzz test is failing here. You need to check if bytes is at least 4 len before indexing it.

Comment (sanket1729) [reply bitcoin/src/psbt/serialize.rs:71]: To reproduce fuzz failures, we you can look at the offending hex. "62" in this case https://github.com/rust-bitcoin/rust-bitcoin/actions/runs/3659817137/jobs/6186247711#step:6:110

and paste in `duplicate_crash` function of failing fuzz test (`deserialize_psbt` in this case). run `cargo test` to debug more.

**Comment by DanGould** [comment]
@sanket1729 Thank you for clarifying. I didn't realize "62" was the fuzz input sitting by its lonesome in console output there.

I made one more ammendment in this recent push, using `crate::io::Cursor` instead of finding the right `Cursor` switching  based on `cfg` features:

https://github.com/rust-bitcoin/rust-bitcoin/blob/27362c10f51682ad2a64312413c787195f3346a0/bitcoin/src/psbt/macros.rs#L55-L60

**Comment by tcharding** [comment]
Guess I was wrong :)

**Review by tcharding — DISMISSED** [review]
ACK 27362c10f51682ad2a64312413c787195f3346a0

File: bitcoin/src/psbt/serialize.rs
Comment (apoelstra) [inline bitcoin/src/psbt/serialize.rs]: In 1177d562a719fb183147630fd627f8955f14289e:

trailing whitespace

**Comment by apoelstra** [comment]
In  1b64e702aa838a54f83f5a2267657d2c41d8e8e9

`cargo +nightly build --features=base64` fails with
```

error[E0277]: the trait bound `PartiallySignedTransaction: Encodable` is not satisfied
   --> bitcoin/src/psbt/mod.rs:793:75
    |
793 |             write!(f, "{}", Base64Display::with_config(&encode::serialize(self), base64::STANDARD))
    |                                                         ----------------- ^^^^ the trait `Encodable` is not implemented for `PartiallySignedTransaction`
    |                                                         |
    |                                                         required by a bound introduced by this call
    |
    = help: the following other types implement trait `Encodable`:
              &'a T
              &'a mut T
              (T0, T1)
              (T0, T1, T2)
              (T0, T1, T2, T3)
              (T0, T1, T2, T3, T4)
              (T0, T1, T2, T3, T4, T5)
              (T0, T1, T2, T3, T4, T5, T6)
            and 107 others
note: required by a bound in `serialize`
   --> bitcoin/src/consensus/encode.rs:118:21
    |
118 | pub fn serialize<T: Encodable + ?Sized>(data: &T) -> Vec<u8> {
    |                     ^^^^^^^^^ required by this bound in `serialize`

error[E0277]: the trait bound `PartiallySignedTransaction: Decodable` is not satisfied
   --> bitcoin/src/psbt/mod.rs:803:13
    |
803 |             encode::deserialize(&data).map_err(PsbtParseError::PsbtEncoding)
    |             ^^^^^^^^^^^^^^^^^^^ the trait `Decodable` is not implemented for `PartiallySignedTransaction`
    |
    = help: the following other types implement trait `Decodable`:
              (T0, T1)
              (T0, T1, T2)
              (T0, T1, T2, T3)
              (T0, T1, T2, T3, T4)
              (T0, T1, T2, T3, T4, T5)
              (T0, T1, T2, T3, T4, T5, T6)
              (T0, T1, T2, T3, T4, T5, T6, T7)
              AddrV2
            and 101 others
note: required by a bound in `deserialize`
   --> bitcoin/src/consensus/encode.rs:132:23
    |
132 | pub fn deserialize<T: Decodable>(data: &[u8]) -> Result<T, Error> {
    |                       ^^^^^^^^^ required by this bound in `deserialize`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `bitcoin` due to 2 previous errors
```

**Review by apoelstra — APPROVED** [review]
ACK c4363e5ba1880a2f815edcfcad1b620805c798b4

**Comment by DanGould** [comment]
I combed over each commit ensuring they contain only my best work. Thank you very much for the thorough reviews.

**Review by tcharding — APPROVED** [review]
ACK c4363e5ba1880a2f815edcfcad1b620805c798b4

**Comment by tcharding** [comment]
Thanks for sticking with it @DanGould!

**Review by sanket1729 — APPROVED** [review]
ACK 1b7b08aa5d08e5d2c19c67084d2530b41a8bb311

File: bitcoin/src/psbt/map/global.rs
Comment (sanket1729) [inline bitcoin/src/psbt/map/global.rs:38]: These should be psbt::Serialize, not encode::Serialize.

**Review by sanket1729 — APPROVED** [review]
ACK c4363e5ba1880a2f815edcfcad1b620805c798b4. One small comment, but can be addressed in follow up if there is nothing else from other reviewers.

**Review by apoelstra — APPROVED** [review]
ACK c4363e5ba1880a2f815edcfcad1b620805c798b4

**Comment by apoelstra** [comment]
@Kixunil it looks like you're still blocking this merge -- should we remove the `Psbt` variant from `consensus::encode::Error` in this PR or can it wait for a later one?

**Comment by DanGould** [comment]
I have that drafted `Psbt` error variant removal for the next PR. Hoping it can wait

**Comment by apoelstra** [comment]
Sounds good to me -- but I'll wait for Kix to comment before merging.

**Comment by Kixunil** [comment]
Sorry, didn't have much time. I'll look into this soon.

File: bitcoin/tests/psbt.rs
Comment (Kixunil) [inline bitcoin/tests/psbt.rs]: I know this is in tests but it still takes reviewer time to realize that. I'd prefer `redeem_script.as_ref().unwrap().as_bytes()`

File: bitcoin/src/psbt/macros.rs
Comment (Kixunil) [inline bitcoin/src/psbt/macros.rs:57]: `Cursor` shouldn't be needed, `&[u8]` already impls `Read`

File: bitcoin/src/psbt/raw.rs
Comment (Kixunil) [inline bitcoin/src/psbt/raw.rs:129]: Again, shouldn't be needed.

**Review by Kixunil — APPROVED** [review]
ACK c4363e5ba1880a2f815edcfcad1b620805c798b4

None of the issues I found is serious, they can be resolved in future PRs, including the error.

**Comment by apoelstra** [comment]
@DanGould let me know if you'd like to address @Kixunil's nits, otherwise we can merge this.

**Comment by DanGould** [comment]
please merge

File: bitcoin/src/psbt/raw.rs
Comment (DanGould) [reply bitcoin/src/psbt/raw.rs:129]: I plan to address all of these in the following PR

File: bitcoin/tests/psbt.rs
Comment (DanGould) [reply bitcoin/tests/psbt.rs]: May you share more about this preference? Is this just to avoid the clone()?

File: bitcoin/src/psbt/serialize.rs
Comment (DanGould) [reply bitcoin/src/psbt/serialize.rs]: This was not necessary and removed in a later commit.

File: bitcoin/src/psbt/map/global.rs
Comment (DanGould) [reply bitcoin/src/psbt/map/global.rs:38]: Should they be? None of `version: i32`, `input`, `output`, `lock_time` are `psbt::map::Serialize` yet.

BIP 174 spec PSBT_GLOBAL_UNSIGNED_TX <valuedata> = "The transaction in [consensus] network serialization. The scriptSigs and witnesses for each input must be empty. The transaction must be in the old serialization format (without witnesses)."

Fortunately PSBTv2 does away with this hack that nests the consensus raw tx map within the psbt map and gives everything relevant in psbt a proper field.

File: bitcoin/tests/psbt.rs
Comment (tcharding) [reply bitcoin/tests/psbt.rs]: I'll take a guess at what he means :) The `clone` causes reviewers to stop and think "why is there a clone", then they have to think more and see that its in test code so its ok to clone. Based on the idea that we should always optimize for reviewers (because review time is always in short supply) it would be better if there was no `clone`.

How'd I go @Kixunil?

Comment (tcharding) [reply bitcoin/tests/psbt.rs]: And ironically that long response probably wastes reviewers time :)

File: bitcoin/src/psbt/serialize.rs
Comment (elichai) [reply bitcoin/src/psbt/serialize.rs:53]: They actually call the same underlying function:
extend: https://doc.rust-lang.org/src/alloc/vec/mod.rs.html#2987
extend_from_slice: https://doc.rust-lang.org/src/alloc/vec/mod.rs.html#2392

File: bitcoin/tests/psbt.rs
Comment (DanGould) [reply bitcoin/tests/psbt.rs]: and now I won't waste it in the future 😄

Comment (Kixunil) [reply bitcoin/tests/psbt.rs]: Yes, what @tcharding said and this is amplified by the code pretty obviously not needing it since all other parts work on references.

**Pull request merged**
